File: language_implementations/a_star.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <cmath>
#include <unordered_map>

using namespace std;

struct Node {
    int x, y, g, h;
    bool operator<(const Node& other) const {
        return g + h > other.g + other.h;
    }
};

int manhattan(int x1, int y1, int x2, int y2) {
    return abs(x1 - x2) + abs(y1 - y2);
}

vector<pair<int, int>> neighbors = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

bool isValid(int x, int y, const vector<vector<int>>& grid) {
    return x >= 0 && y >= 0 && x < grid.size() && y < grid[0].size() && grid[x][y] == 0;
}

//reference: astar function from chatgpt
vector<pair<int, int>> aStar(vector<vector<int>>& grid, pair<int, int> start, pair<int, int> goal) {
    priority_queue<Node> openSet;
    unordered_map<int, int> cameFrom;
    openSet.push({start.first, start.second, 0, manhattan(start.first, start.second, goal.first, goal.second)});

    while (!openSet.empty()) {
        Node current = openSet.top();
        openSet.pop();
        if (current.x == goal.first && current.y == goal.second) {
            vector<pair<int, int>> path;
            while (current.g != 0) {
                path.push_back({current.x, current.y});
                int key = current.x * grid[0].size() + current.y;
                current.x = cameFrom[key] / grid[0].size();
                current.y = cameFrom[key] % grid[0].size();
            }
            path.push_back(start);
            reverse(path.begin(), path.end());
            return path;
        }

        for (auto [dx, dy] : neighbors) {
            int nx = current.x + dx, ny = current.y + dy;
            if (isValid(nx, ny, grid)) {
                int new_g = current.g + 1;
                int h = manhattan(nx, ny, goal.first, goal.second);
                openSet.push({nx, ny, new_g, h});
                cameFrom[nx * grid[0].size() + ny] = current.x * grid[0].size() + current.y;
                grid[nx][ny] = 1;
            }
        }
    }
    return {};
}

int main() {
    vector<vector<int>> grid = {
        {0, 0, 0, 0, 0},
        {0, 1, 1, 1, 0},
        {0, 0, 0, 0, 0},
        {0, 1, 1, 1, 0},
        {0, 0, 0, 0, 0}};
    pair<int, int> start = {0, 0}, goal = {4, 4};

    auto path = aStar(grid, start, goal);
    for (auto [x, y] : path) {
        cout << "(" << x << ", " << y << ") ";
    }
    return 0;
}

File: language_implementations/a_star.rs
use std::collections::BinaryHeap;
use std::cmp::Ordering;

#[derive(Eq, PartialEq)]
struct Node {
    x: usize,
    y: usize,
    g: usize,
    h: usize,
    parent: Option<Box<Node>>,
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> Ordering {
        (other.g + other.h).cmp(&(self.g + self.h))
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn manhattan(x1: usize, y1: usize, x2: usize, y2: usize) -> usize {
    ((x1 as isize - x2 as isize).abs() + (y1 as isize - y2 as isize).abs()) as usize
}

fn is_valid(x: usize, y: usize, grid: &Vec<Vec<i32>>) -> bool {
    x < grid.len() && y < grid[0].len() && grid[x][y] == 0
}

fn reconstruct_path(mut current: &Node) -> Vec<(usize, usize)> {
    let mut path = Vec::new();
    while let Some(parent) = &current.parent {
        path.push((current.x, current.y));
        current = parent;
    }
    path.push((current.x, current.y));
    path.reverse();
    path
}

fn a_star(grid: &mut Vec<Vec<i32>>, start: (usize, usize), goal: (usize, usize)) -> Vec<(usize, usize)> {
    let mut open_set = BinaryHeap::new();
    open_set.push(Node {
        x: start.0,
        y: start.1,
        g: 0,
        h: manhattan(start.0, start.1, goal.0, goal.1),
        parent: None,
    });

    while let Some(current) = open_set.pop() {
        if (current.x, current.y) == goal {
            return reconstruct_path(&current);
        }

        let neighbors = [(1, 0), (-1, 0), (0, 1), (0, -1)];
        for (dx, dy) in neighbors {
            let nx = current.x as isize + dx;
            let ny = current.y as isize + dy;

            if nx >= 0 && ny >= 0 {
                let nx = nx as usize;
                let ny = ny as usize;

                if is_valid(nx, ny, grid) {
                    open_set.push(Node {
                        x: nx,
                        y: ny,
                        g: current.g + 1,
                        h: manhattan(nx, ny, goal.0, goal.1),
                        parent: Some(Box::new(current.clone())),
                    });
                    grid[nx][ny] = 1;
                }
            }
        }
    }
    Vec::new() // No path found
}

fn main() {
    let mut grid = vec![
        vec![0, 0, 0, 0, 0],
        vec![0, 1, 1, 1, 0],
        vec![0, 0, 0, 0, 0],
        vec![0, 1, 1, 1, 0],
        vec![0, 0, 0, 0, 0],
    ];
    let start = (0, 0);
    let goal = (4, 4);

    let path = a_star(&mut grid, start, goal);
    for (x, y) in path {
        println!("({}, {})", x, y);
    }
}
